src/wallet/rpcwallet.cpp

```diff
@@ -5028,39 +5029,38 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
     int nMinDepth = parseMinconf(DEFAULT_NOTE_CONFIRMATIONS, params, 2, std::nullopt);
 
     // Fee in Zatoshis, not currency format)
-    CAmount nFee = DEFAULT_FEE;
-    if (params.size() > 3) {
-        if (params[3].get_real() == 0.0) {
-            nFee = 0;
-        } else {
-            nFee = AmountFromValue( params[3] );
-        }
+    std::optional<CAmount> nFee;
+    if (params.size() > 3 && params[3].get_real() != -1.0) {
+        CAmount fixedFee = AmountFromValue( params[3] );
 
         // Check that the user specified fee is not absurd.
         // This allows amount=0 (and all amount < DEFAULT_FEE) transactions to use the default network fee
         // or anything less than DEFAULT_FEE instead of being forced to use a custom fee and leak metadata
         if (nTotalOut < DEFAULT_FEE) {
-            if (nFee > DEFAULT_FEE) {
+            if (fixedFee > DEFAULT_FEE) {
                 throw JSONRPCError(
                         RPC_INVALID_PARAMETER,
-                        strprintf("Small transaction amount %s has fee %s that is greater than the default fee %s", FormatMoney(nTotalOut), FormatMoney(nFee), FormatMoney(DEFAULT_FEE)));
+                        strprintf("Small transaction amount %s has fee %s that is greater than the default fee %s", FormatMoney(nTotalOut), FormatMoney(fixedFee), FormatMoney(DEFAULT_FEE)));
             }
         } else {
             // Check that the user specified fee is not absurd.
-            if (nFee > nTotalOut) {
+            if (fixedFee > nTotalOut) {
                 throw JSONRPCError(
                         RPC_INVALID_PARAMETER,
-                        strprintf("Fee %s is greater than the sum of outputs %s and also greater than the default fee", FormatMoney(nFee), FormatMoney(nTotalOut)));
+                        strprintf("Fee %s is greater than the sum of outputs %s and also greater than the default fee", FormatMoney(fixedFee), FormatMoney(nTotalOut)));
             }
         }
+        nFee = fixedFee;
     }
 
     // Use input parameters as the optional context info to be returned by z_getoperationstatus and z_getoperationresult.
     UniValue o(UniValue::VOBJ);
     o.pushKV("fromaddress", params[0]);
     o.pushKV("amounts", params[1]);
     o.pushKV("minconf", nMinDepth);
-    o.pushKV("fee", std::stod(FormatMoney(nFee)));
+    if (nFee.has_value()) {
+        o.pushKV("fee", std::stod(FormatMoney(nFee.value())));
```

Comment:
I don't understand why we're formatting the value and then re-parsing it.

Reply:
My guess is that this is for decimal handling: `FormatMoney` formats a zatoshi value as a ZEC string (using fixed-point decimal), and then `stod` is being used to convert _that_ to a floating-point decimal.

Reply:
I’m a bit confused, actually. I was thinking what @str4d said, but I looked at other code, the `z_gettotalbalance` docs claim
```
Result:
{
  "transparent": xxxxx,     (numeric) the total balance of transparent funds
  "private": xxxxx,         (numeric) the total balance of shielded funds (in all shielded addresses)
  "total": xxxxx,           (numeric) the total balance of both transparent and shielded funds
}
```
but the code does
```c++
result.pushKV("transparent", FormatMoney(nBalance));
result.pushKV("private", FormatMoney(nPrivateBalance));
result.pushKV("total", FormatMoney(nTotalBalance));
```
The z_sendmany fee entry is the only use of `stod` in the repo. Other places use `ValueFromAmount`, which does more manual CAmount->std::string->Univalue::VNUM conversion.

So, this particular case should probably be changed to use `ValueFromAmount`, and I wonder if we need to audit the rest of the RPC API to fix things like the snippet above where we are using strings instead of numbers in the JSON?

Reply:
Also wondering if `ValueFromAmount` should be changed to use `FormatMoney` as its stringifier (not in this PR), or if it should be considered a coincidence that the human-readable format and the en-route-to-JSON format are the same.

Reply:
I checked what `ValueFromAmount` is doing and it is definitely the right thing to use here.